### PR TITLE
Optimize syntactic classification walk

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Drops us about 0.7% during scrolling from:

image

to

image

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Classification/Worker.cs [hunk 1]
```diff
@@ -80,10 +80,28 @@ private void ClassifyNodeOrToken(SyntaxNodeOrToken nodeOrToken)
...
+        while (stack.TryPop(out var current))
         {
             _cancellationToken.ThrowIfCancellationRequested();
-            ClassifyToken(token);
+
+            // It's ok that we're not pushing in reverse.  The caller (TotalClassificationTaggerProvider) will be
+            // sorting the results before doing anything with them.
+            foreach (var child in current.ChildNodesAndTokens())
+            {
+                if (child.IsNode)
```
Comment (ToddGrun) [inline src/Workspaces/CSharp/Portable/Classification/Worker.cs:94]: >if (child.IsNode) [](http://example.com/codeflow?start=16&length=17)

nit question: I've seen this in a couple places, why not just use the "bool AsNode(out SyntaxNode? node)" method?

Comment (sharwell) [reply src/Workspaces/CSharp/Portable/Classification/Worker.cs:94]: If possible, I would prefer it be called `IsNode(out SyntaxNode?)`. I don't think this is possible with a property and method in the same type, but may be possible with it defined as an extension method.

File: src/Workspaces/CSharp/Portable/Classification/Worker.cs [hunk 2]
```diff
@@ -80,10 +80,28 @@ private void ClassifyNodeOrToken(SyntaxNodeOrToken nodeOrToken)
...
+        stack.Push(node);
+
+        while (stack.TryPop(out var current))
         {
             _cancellationToken.ThrowIfCancellationRequested();
-            ClassifyToken(token);
+
+            // It's ok that we're not pushing in reverse.  The caller (TotalClassificationTaggerProvider) will be
+            // sorting the results before doing anything with them.
+            foreach (var child in current.ChildNodesAndTokens())
```
Comment (ToddGrun) [inline src/Workspaces/CSharp/Portable/Classification/Worker.cs:92]: >ChildNodesAndTokens [](http://example.com/codeflow?start=42&length=19)

this is way off topic, but I took another look at the SyntaxList.Enumerator's call to ItemInternal, and I'm not clear why the slotIndex isn't held between calls. By not doing so, the first loop in ItemInternal would need to walk the green slots until it reaches the appropriate slot index.

Again, off topic, so nothing needed in this PR, but it seems odd.

File: src/Workspaces/CSharp/Portable/Classification/Worker.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Classification/Worker.cs:94]: I didn't know the AsNode method existed. Will update to that.

File: src/Workspaces/CSharp/Portable/Classification/Worker.cs [hunk 3]
```diff
@@ -80,10 +80,28 @@ private void ClassifyNodeOrToken(SyntaxNodeOrToken nodeOrToken)
...
+            {
+                if (child.IsNode)
+                {
+                    var childNode = child.AsNode()!;
+                    if (childNode.FullSpan.IntersectsWith(_textSpan))
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Classification/Worker.cs:97]: I'll def try that

File: src/Workspaces/CSharp/Portable/Classification/Worker.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Classification/Worker.cs:91]: def look into that.  there are some comments explaining things.  but it's possible they're stale, or there is room for improvement here.

File: src/Workspaces/CSharp/Portable/Classification/Worker.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Classification/Worker.cs]: done

File: src/Workspaces/CSharp/Portable/Classification/Worker.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Classification/Worker.cs]: done.

File: src/Workspaces/CSharp/Portable/Classification/Worker.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Classification/Worker.cs:96]: Note: i'm saying @ToddGrun should look into this :)

**Pull request merged**